### PR TITLE
Removed broiler code

### PR DESCRIPTION
**Pull request by faizanvelotio**
Today most of my time went on reading the Typescript handbook, refreshing React concepts and deploying on Heroku

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,42 @@
-import React from 'react';
-import logo from './logo.svg';
-import './App.css';
+import { Route, Switch } from "react-router-dom";
+
+import Users from "./pages/Users";
```
Comment (harshilvelotio) [inline src/App.tsx]: Let's use absolute imports

Comment (harshilvelotio) [inline src/App.tsx]:
```suggestion
import Users from "src/pages/Users";
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,42 @@
...
+import UserPosts from "./pages/Posts";
+import Home from "./pages/Home";
+import UserTodos from "./pages/Todos";
+import UserForm from "./pages/UserForm";
+import PostForm from "./pages/PostForm";
+
+import "./App.css";
```
Comment (harshilvelotio) [inline src/App.tsx]: We don't need css imports up til now

File: src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,42 @@
...
-          Learn React
-        </a>
-      </header>
-    </div>
-  );
+    return (
+        <div className="App">
+            <Navbar />
+            <Switch>
+                <Route exact path="/" children={<Home />} />
```
Comment (harshilvelotio) [inline src/App.tsx]: Implement routes through a layout HOC

File: src/context/UserContentContext.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import React, { createContext, useReducer } from "react";
+import { User, PostWithComment, Todos, Comment } from "../interfaces";
+
+interface AppState {
+    currentUser: User | null;
+    user_posts: { userId: number; posts: Array<PostWithComment> | null };
```
Comment (harshilvelotio) [inline src/context/UserContentContext.tsx]:
```suggestion
    user_posts: { userId: number; posts: PostWithComment[] | null };
```

File: src/context/UserContentContext.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    user_todos: { userId: number; todos: Array<Todos> | null };
+}
+
+type Action =
+    | { type: "SET_USER"; payload: User }
```
Comment (harshilvelotio) [inline src/context/UserContentContext.tsx]:
```suggestion
    | { type: UserActions.SET_USER; payload: User }
```

File: src/interfaces.d.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+export interface User {
```
Comment (harshilvelotio) [inline src/interfaces.d.ts]: In a `.d.ts` file you don't need to export, you can just `declare` the interface 
```suggestion
declare interface User {
```

File: src/pages/Posts.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import useLocationId from "../utils/useLocationId";
+// import useApiHook from "../utils/useApiHook";
+import { Post, PostWithComment, ViewPost } from "../interfaces";
+import { UserContentContext } from "../context/UserContentContext";
+import SinglePost from "./SinglePost";
+
+function Posts() {
+    const userId: number = useLocationId(); // Get the id parameter for URL
+    const [isLoading, setIsLoading] = useState(true); // If the posts are loading or not
+    const [viewablePost, setViewablePost] = useState<ViewPost>({
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]: Store the initial state in a separate variable outside the component

Comment (harshilvelotio) [inline src/pages/Posts.tsx]:
```suggestion
    const [viewablePost, setViewablePost] = useState<ViewPost>(initialViewablePost)
```

File: src/pages/Posts.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        index: -1,
+    });
+
+    const { state, dispatch } = useContext(UserContentContext);
+    const { user_posts } = state;
+    useEffect(() => {
+        // If the posts in the state is of different user, then update the state,
+        // else update render with the same state
+        if (userId !== user_posts.userId) {
+            axios
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]: Extract all API calls for posts in a separate `api/posts.ts`
```
export const getUserPosts = (userId: string): Promise<PostData[]> => axios.post<AxiosResponse<PostData[]>>(...
```

And in `Posts.tsx`
```
const Posts: React.FC = () => {

    const fetchUserPosts = useCallback(async (userId: string) => {
      try {
        setLoading(true)
        const resp = await getUserPosts(userId);
        setState(resp)
      } catch (e) {
          toast('Error occured', e)
      } finally {
          setLoading(false)
      }
  }, []);

  useEffect(() => {
      fetchUserPosts(userId)
  }, [userId, fetchUserPosts])
};
```

File: src/pages/Posts.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                .get(`/users/${userId}/posts`)
+                .then((response) => {
+                    // Create an array where each object has a post and an empty comment array
+                    let posts_without_comments = response.data;
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]: This logic should live inside the reducer

File: src/pages/Posts.tsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                    index={viewablePost.index}
+                />
+            )}
+            {isLoading ? (
+                <div>Loading...</div>
+            ) : (
+                <>
+                    <h1 style={{ textAlign: "center" }}>Posts</h1>
+                    {user_posts.posts &&
+                        user_posts.posts.map((val: PostWithComment, idx) => (
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]: Create a separarte function called `renderUserPosts = useCallback(() => user_posts.map ,[])` to handle rendering here

File: src/pages/Posts.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                    setter={setViewablePost}
+                    index={viewablePost.index}
+                />
+            )}
+            {isLoading ? (
+                <div>Loading...</div>
+            ) : (
+                <>
+                    <h1 style={{ textAlign: "center" }}>Posts</h1>
+                    {user_posts.posts &&
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]:
```suggestion
                    {userPosts.posts &&
```

File: src/pages/Posts.tsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                        user_posts.posts.map((val: PostWithComment, idx) => (
+                            <div
+                                key={val.post.id}
+                                onClick={() => {
```
Comment (harshilvelotio) [inline src/pages/Posts.tsx]: Create separate handlers for onClicks and surround them with callbacks as well 
```
const handlePostClick = useCallback((postId) => {}, [])
```

Comment (harshilvelotio) [reply src/pages/Posts.tsx]: Enable `react-hooks/exhaustive-deps` rule in your linter

File: src/pages/SinglePost.tsx [hunk 13]
```diff
@@ -0,0 +1,82 @@
+import { ViewPost, PostWithComment, Comment } from "../interfaces";
+import { UserContentContext } from "../context/UserContentContext";
+import { useContext, useEffect } from "react";
+import axios from "axios";
+
+function SinglePost({
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]:
```suggestion
const SinglePost: React.FC<Props> = ({ post, index }) => {}
```

File: src/pages/SinglePost.tsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    post,
+    index,
+    setter,
+}: {
+    post: PostWithComment;
+    index: number;
+    setter: React.Dispatch<React.SetStateAction<ViewPost>>;
+}) {
+    const closeSinglePostDialog = () => {
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]: Surround with useCallback

File: src/pages/SinglePost.tsx [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        setter((prevState) => ({ ...prevState, view: false }));
+    };
+    const { dispatch } = useContext(UserContentContext);
+
+    useEffect(() => {
+        if (post.comments === null) {
+            axios.get(`/comments?postId=${post.post.id}`).then((response) => {
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]: Use API pattern here as well

File: src/pages/SinglePost.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+                // console.log("from single post axios", response.data);
+                dispatch({
+                    type: "SET_COMMENT_FOR_POST",
+                    payload: { postIndex: index, comments: response.data },
+                });
+            });
+        }
+    }, [dispatch, post, index]);
+
+    // console.log(post, index);
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]: Remove unnecessary console logs

File: src/pages/SinglePost.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+            >
+                Close Dialog
+            </div>
+            <div>Title: {post.post.title}</div>
+            <div>Body: {post.post.body}</div>
+            {post.comments ? (
+                <div>
+                    <h3>Comments</h3>
+
+                    {post.comments.length !== 0 ? (
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]:
```suggestion
                    {post.comments.length ? (
```

File: src/pages/SinglePost.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+                        post.comments.map((comment: Comment) => {
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]: Use a `renderPostComments` function here

File: src/pages/SinglePost.tsx [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+                Single Post
+            </h2>
+            <div
+                onClick={closeSinglePostDialog}
+                style={{ color: "blue", textAlign: "right" }}
+            >
+                Close Dialog
+            </div>
+            <div>Title: {post.post.title}</div>
+            <div>Body: {post.post.body}</div>
```
Comment (harshilvelotio) [inline src/pages/SinglePost.tsx]:
```suggestion
            <div>Body: {post.post.body}</div>
            {renderComments()}
```

File: src/pages/Todos.tsx [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    const [isLoading, setIsLoading] = useState(true);
+
+    const { state, dispatch } = useContext(UserContentContext);
+    const { user_todos } = state;
+
+    useEffect(() => {
+        // If the posts in the state is of different user, then update the state,
+        // else update render with the same state
+        // console.log("from TODOS useeffect");
+        if (userId !== user_todos.userId) {
```
Comment (harshilvelotio) [inline src/pages/Todos.tsx]: Use API pattern here

File: src/pages/Todos.tsx [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+                    <button>Go to Posts</button>
+                </Link>
+            </div>
+            {/* {error && <div>{error.message}</div>} */}
+            {isLoading ? (
+                <div>Loading...</div>
+            ) : (
+                <>
+                    <h1 style={{ textAlign: "center" }}>Todos</h1>
+                    {user_todos.todos &&
```
Comment (harshilvelotio) [inline src/pages/Todos.tsx]: Use `renderTodos` function here

File: src/pages/UserForm.tsx [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+import Input from "./Input";
+import * as Yup from "yup";
+import useLocationId from "../utils/useLocationId";
+
+function UserForm({ edit }: { edit: boolean }) {
+    // const userId: number = useLocationId();
+
+    // console.log("user id", userId);
+
+    const validate = Yup.object({
```
Comment (harshilvelotio) [inline src/pages/UserForm.tsx]: Validation schema can live outside of the component

File: src/pages/UserForm.tsx [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+            new RegExp(/[\d -]+/),
+            "Phone number should be numeric"
+        ),
+        zipcode: Yup.string().matches(
+            new RegExp(/\d\d\d\d\d\d/),
+            "Zipcode should be of length 6."
+        ),
+    });
+    return (
+        <Formik
```
Comment (harshilvelotio) [inline src/pages/UserForm.tsx]: Use `useFormik` hook instead

File: src/pages/UserForm.tsx [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+            initialValues={{
```
Comment (harshilvelotio) [inline src/pages/UserForm.tsx]: Initial values can be a separate variable outside component

File: src/pages/UserForm.tsx (hunk 23 above)
Comment (harshilvelotio) [reply src/pages/UserForm.tsx]: Pass the ValueType as a generic to the `useFormik<FormType>`

File: src/pages/Users.tsx [hunk 25]
```diff
@@ -0,0 +1,38 @@
+// import { useEffect } from "react";
+import { useHistory } from "react-router-dom";
+
+import useApiHook from "../utils/useApiHook";
+import { User } from "../interfaces";
+
+function Users() {
+    const { data, isPending, error } = useApiHook<User>("/users");
```
Comment (harshilvelotio) [inline src/pages/Users.tsx]: Use an effect hook with Axios here instead

File: src/App.tsx (hunk 2 above)
Comment (faizanvelotio) [reply src/App.tsx]: I cannot understand this comment, so do I need to remove CSS imports?

File: src/App.tsx [hunk 26]
```diff
@@ -1,24 +1,33 @@
...
-import './App.css';
+import { Route, Switch } from "react-router-dom";
+
+import Users from "src/pages/Users";
+import UserPosts from "src/pages/Posts";
+import Home from "src/pages/Home";
+import UserTodos from "src/pages/Todos";
+import UserForm from "src/pages/UserForm";
+import PostForm from "src/pages/PostForm";
+import Layout from "./Layout";
```
Comment (harshilvelotio) [inline src/App.tsx]: Use absolute imports

File: src/App.tsx [hunk 27]
```diff
@@ -1,24 +1,33 @@
...
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+      <Switch>
+        <Route
```
Comment (harshilvelotio) [inline src/App.tsx]:
```suggestion
        <Layout header={false} ....
```

File: src/Layout.tsx [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+
+interface LayoutProps {
+  header?: boolean;
+  component: React.FC;
+}
+
+const Layout: React.FC<LayoutProps> = ({ header, component, ...rest }) => {
+  const Component: React.FC = component;
+  return (
+    <div className="main-container">
```
Comment (harshilvelotio) [inline src/Layout.tsx]:
```suggestion
    return <Route render={(props) => <><div></div></>}
```

File: src/api/Post.ts [hunk 29]
```diff
@@ -0,0 +1,8 @@
+import axios, { AxiosResponse } from "axios";
+
+export const getUserPosts = (userId: number): Promise<AxiosResponse<Post[]>> =>
+  axios.get(`/users/${userId}/posts`);
```
Comment (harshilvelotio) [inline src/api/Post.ts]:
```suggestion
  axios.get<Post[]>(`/users/${userId}/posts`);
```

File: src/api/axiosConfig.ts [hunk 30]
```diff
@@ -0,0 +1,7 @@
+import axios from "axios";
+
+const axiosConfig = axios.create({
+  baseURL: "https://jsonplaceholder.typicode.com/",
```
Comment (harshilvelotio) [inline src/api/axiosConfig.ts]: use a `.env.development` or a `.env.production` file instead

Comment (harshilvelotio) [inline src/api/axiosConfig.ts]: use a `.env.development` or a `.env.production` file instead

File: src/components/CommentCard.tsx [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+            fontSize: "1.2rem",
+            lineHeight: "1.5rem",
+          }}
+        >
+          {comment.name}
+        </Box>
+        <Typography
+          variant="caption"
+          component="div"
+          sx={{ color: "#A8A8A8", margin: "0.15rem 0" }}
```
Comment (harshilvelotio) [inline src/components/CommentCard.tsx]: Never use naked hex values
All colours should come from the theme

Comment (harshilvelotio) [inline src/components/CommentCard.tsx]: Never use naked hex values
All colours should come from the theme

Comment (harshilvelotio) [reply src/components/CommentCard.tsx]: ALWAYS use `mt`, `mx`, `my`, etc for layout designing

Comment (harshilvelotio) [reply src/components/CommentCard.tsx]: ALWAYS use `mt`, `mx`, `my`, etc for layout designing

File: src/components/CommentCard.tsx [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+        >
+          {comment.email}
+        </Typography>
+        <Box
+          sx={{
+            fontSize: "0.8rem",
```
Comment (harshilvelotio) [inline src/components/CommentCard.tsx]: source of theming always from a single place

Comment (harshilvelotio) [inline src/components/CommentCard.tsx]: source of theming always from a single place

**Review by harshilvelotio — CHANGES_REQUESTED** [review]

**Pull request merged**